### PR TITLE
Add user:modify command to core

### PR DESCRIPTION
**Pull request by sharidas**
Add user:modify command to core.
This command will help user to set the
email and displayname using occ command.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Intention of adding this command is to provide admins the ability to set the email address and displayname for users using occ command.

## Related Issue

https://github.com/owncloud/core/issues/30529

## Motivation and Context

This change would help administrators to set the displayname and email address for users using occ command.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: core/Command/User/Modify.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+		$user = $this->userManager->get($uid);
+
+		if (($key === 'email') && ($input->hasParameterOption('--value')) && ($value !== null)) {
+			$user->setEMailAddress($value);
+			return 0;
+		}
+
+		if (($key === 'displayname') && ($input->hasParameterOption('--value')) && ($value !== null)) {
+			if ($user->canChangeDisplayName() === false) {
```
Comment (sharidas) [inline core/Command/User/Modify.php]: Argh doesn't make sense, as the administrator triggers the command.

File: core/Command/User/Modify.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+			return 1;
+		}
+
+		$user = $this->userManager->get($uid);
+
+		if (($key === 'email') && ($input->hasParameterOption('--value')) && ($value !== null)) {
+			$user->setEMailAddress($value);
+			return 0;
+		}
+		if (($key === 'email') && ($value === null)) {
```
Comment (sharidas) [inline core/Command/User/Modify.php]: Lets say if user doesn't want to modify the settings `displayname` or `email` then lets display it and return. Same applies for https://github.com/owncloud/core/pull/30540/files#diff-bb62ccd405c6abbc168c61fb1ea08817R115. So basically --value=something means we are setting it.

File: core/Command/User/Modify.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+use OC\Core\Command\Base;
+use OCP\IUserManager;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class Modify extends Base {
+
+	private $userManager;
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: phpdoc missing

File: core/Command/User/Modify.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	private $allowedKeys = ['email', 'displayname'];
+
+	public function __construct(IUserManager $userManager) {
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: phpdoc missing

File: core/Command/User/Modify.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			->setDescription('Modify user details')
+			->addArgument(
+				'uid',
+				InputArgument::REQUIRED,
+				'User ID used to login'
+			)
+			->addArgument(
+				'key',
+				InputArgument::OPTIONAL,
+				'Setting key to set, get or delete',
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: get? I'd not expect a get option in a modify command

File: core/Command/User/Modify.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	protected function validateInput(InputInterface $input) {
+		$uid = $input->getArgument('uid');
+		if ($this->userManager->userExists($uid) === false) {
+			throw new \InvalidArgumentException("The user $uid does not exist");
+		}
+
+		if (($input->getArgument('key') === '') or ($input->getArgument('key') === null)) {
+			throw new \InvalidArgumentException("The key cannot be empty");
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: single quote

File: core/Command/User/Modify.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+			);
+	}
+
+	protected function validateInput(InputInterface $input) {
+		$uid = $input->getArgument('uid');
+		if ($this->userManager->userExists($uid) === false) {
+			throw new \InvalidArgumentException("The user $uid does not exist");
+		}
+
+		if (($input->getArgument('key') === '') or ($input->getArgument('key') === null)) {
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: use || not or

File: core/Command/User/Modify.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			throw new \InvalidArgumentException("The key cannot be empty");
+		}
+
+		if (in_array($input->getArgument('key'), $this->allowedKeys) === false) {
+			throw new \InvalidArgumentException("Supported keys are " . implode(', ', $this->allowedKeys));
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: single quote

File: tests/Core/Command/User/ModifyTest.php [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	protected function setUp() {
+		parent::setUp();
+
+		$this->userManager = $this->createMock(IUserManager::class);
+		$this->consoleInput = $this->createMock(InputInterface::class);
+		$this->consoleOutput = $this->createMock(OutputInterface::class);
+	}
+
+	public function getCommand(array $methods = []) {
+		$mock = $this->getMockBuilder('OC\Core\Command\User\Modify')
```
Comment (DeepDiver1975) [inline tests/Core/Command/User/ModifyTest.php]: Modify::class

File: tests/Core/Command/User/ModifyTest.php [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+			->setConstructorArgs([$this->userManager])
+			->setMethods($methods)
+			->getMock();
+		return $mock;
+	}
+
+	public function dataInput() {
+		return [
+			[false, 'foo', '', "The user foo does not exist"],
```
Comment (DeepDiver1975) [inline tests/Core/Command/User/ModifyTest.php]: single quote

File: tests/Core/Command/User/ModifyTest.php [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+			[true, 'foo', null, "The key cannot be empty"],
+			[true, 'foo', 'test', "Supported keys are displayname, email"],
+			[true, 'foo', 'displayname', false],
+		];
+	}
+
+	/**
+	 * @dataProvider dataInput
+	 * @param $uidExist
```
Comment (DeepDiver1975) [inline tests/Core/Command/User/ModifyTest.php]: missing types

File: tests/Core/Command/User/ModifyTest.php [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+			$this->invokePrivate($usermodify, 'validateInput', [$this->consoleInput]);
+			$this->assertFalse($expectedException);
+		} catch (\InvalidArgumentException $e) {
+			$this->assertEquals($expectedException, $e->getMessage());
+		}
+	}
+
+	public function dataExecute() {
+		/*return [
```
Comment (DeepDiver1975) [inline tests/Core/Command/User/ModifyTest.php]: remove?

File: tests/Core/Command/User/ModifyTest.php [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+			$property->setValue($usermodify, ["displayname", "email"]);
+
+			$this->invokePrivate($usermodify, 'validateInput', [$this->consoleInput]);
+			$this->assertFalse($expectedException);
+		} catch (\InvalidArgumentException $e) {
+			$this->assertEquals($expectedException, $e->getMessage());
+		}
+	}
+
+	public function dataExecute() {
```
Comment (DeepDiver1975) [inline tests/Core/Command/User/ModifyTest.php]: I usually start provider methods with provides ...

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: core/Command/User/Modify.php [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+			$this->validateInput($input);
+		} catch (\InvalidArgumentException $e) {
+			$output->writeln('<error>' . $e->getMessage() . '</error>');
+			return 1;
+		}
+
+		$user = $this->userManager->get($uid);
+
+		if (($key === 'email') && ($input->hasParameterOption('--value')) && ($value !== null)) {
+			if ($this->mailer->validateMailAddress($value) === true) {
```
Comment (sharidas) [inline core/Command/User/Modify.php]: Additional check to verify invalid address provided.

File: core/Command/User/Modify.php [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+			->setDescription('Modify user details')
+			->addArgument(
+				'uid',
+				InputArgument::REQUIRED,
+				'User ID used to login'
+			)
+			->addArgument(
+				'key',
+				InputArgument::OPTIONAL,
+				'Setting key to set',
```
Comment (mmattel) [inline core/Command/User/Modify.php]: 1. InputArgument::OPTIONAL --> InputArgument::REQUIRED (see line 87)
2. ``Setting key to set`` --> ``Key to be changed``

File: core/Command/User/Modify.php [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+				''
+			)
+			->addOption(
+				'value',
+				null,
+				InputOption::VALUE_REQUIRED,
+				'The new value of the setting.'
```
Comment (mmattel) [inline core/Command/User/Modify.php]: ``The new value of the setting`` --> `` The new value of the key``

File: core/Command/User/Modify.php (hunk 15 above)
Comment (tomneedham) [reply core/Command/User/Modify.php]: Maybe we also give a list of valid values here?

File: core/Command/User/Modify.php [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+			->setName('user:modify')
+			->setDescription('Modify user details')
+			->addArgument(
+				'uid',
+				InputArgument::REQUIRED,
+				'User ID used to login'
+			)
+			->addArgument(
+				'key',
+				InputArgument::OPTIONAL,
```
Comment (mmattel) [inline core/Command/User/Modify.php]: I need to ask again,
why do you have a argument optional when you then throw an excpetion if it is null or empty?
See the query in line 87

**Comment by mmattel** [comment]
LGTM
Please add (and link) a documentation PR to describe the new command.

File: core/Command/User/Modify.php [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+			$output->writeln('<error>' . $e->getMessage() . '</error>');
+			return 1;
+		}
+
+		$user = $this->userManager->get($uid);
+
+		if (($key === 'email') && ($input->hasParameterOption('--value')) && ($value !== null)) {
+			if ($this->mailer->validateMailAddress($value) === true) {
+				$user->setEMailAddress($value);
+				$output->writeln('The email address of ' . $uid . ' updated to ' . $value);
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]:
```php
"The email address of $uid updated to $value"

File: core/Command/User/Modify.php [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+			$output->writeln('<error>Email address provided is invalid</error>');
+			return 1;
+		}
+
+		if (($key === 'displayname') && ($input->hasParameterOption('--value')) && ($value !== null)) {
+			$user->setDisplayName($value);
+			$output->writeln('The displayname of ' . $uid . ' updated to ' . $value);
+			return 0;
+		}
+	}
```
Comment (DeepDiver1975) [inline core/Command/User/Modify.php]: no return at the end?

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by sharidas** [comment]
Updated the PR with the changes requested.

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]
Running ./occ u:m admin displayname has no output - nobody knows whats going on.

And why is value an option?

**Comment by sharidas** [comment]
> And why is value an option?

Modified value from option to argument and marked as required.

File: core/Command/User/Modify.php [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+			return 1;
+		}
+
+		$user = $this->userManager->get($uid);
+		if ($user === null) {
+			$output->writeln('<error>The user ' . $uid . ' does not exist </error>');
+			return 1;
+		}
+
+		if (($key === 'email') && ($value !== null)) {
```
Comment (PVince81) [inline core/Command/User/Modify.php]: is it possible to set the email and display name to empty string ?

File: tests/Core/Command/User/ModifyTest.php [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+
+		$retVal = static::invokePrivate($userModify, 'execute', [$this->consoleInput, $this->consoleOutput]);
+		$this->assertEquals(1, $retVal);
+	}
+
+	public function providerExecuteMethod() {
+		return [
+			['foo', 'displayname', 'foo123'],
+			['foo', 'email', 'foo@bar.com'],
+			['foo', 'email', 'foo@bar.com@foobar'], //Test for invalid email
```
Comment (PVince81) [inline tests/Core/Command/User/ModifyTest.php]: please add tests for setting empty values

File: core/Command/User/Modify.php (hunk 20 above)
Comment (mmattel) [reply core/Command/User/Modify.php]: Imho there must be a username but there may be an eMail address, or can the username also be empty?

If the username can be empty, then we need to check that the eMail address is either already existent or populated by the key.

File: core/Command/User/Modify.php [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+			->setDescription('Modify user details')
+			->addArgument(
+				'uid',
+				InputArgument::REQUIRED,
+				'User ID used to login'
+			)
+			->addArgument(
+				'key',
+				InputArgument::REQUIRED,
+				'Key to be changed. The valid keys are : displayname, email'
```
[the remark this answers is not in the document]
Comment (sharidas) [reply core/Command/User/Modify.php]: Addressed in this version of PR.

File: tests/Core/Command/User/ModifyTest.php (hunk 21 above)
Comment (sharidas) [reply tests/Core/Command/User/ModifyTest.php]: Added tests for empty values. In fact added more cases for input uid/key/values.

File: core/Command/User/Modify.php (hunk 20 above)
Comment (sharidas) [reply core/Command/User/Modify.php]: IMHO lets make it simple, if the username is empty then lets not do anything ( of course display valid message ).

File: tests/Core/Command/User/ModifyTest.php [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+		$this->userManager = $this->createMock(IUserManager::class);
+		$this->mailer = $this->createMock(IMailer::class);
+		$this->consoleInput = $this->createMock(InputInterface::class);
+		$this->consoleOutput = $this->createMock(OutputInterface::class);
+	}
+
+	public function providerValidateInputMethod() {
+		return [
+			[false, 'foo', '', '', 'The user foo does not exist'],
+			[true, '', '', '', 'The user  does not exist'],
```
Comment (sharidas) [inline tests/Core/Command/User/ModifyTest.php]: When the user is empty/null then just provide this message shown in the last argument.

File: tests/Core/Command/User/ModifyTest.php [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+			[true, null, '', '', 'The user  does not exist'],
+			[true, 'foo', null, '', 'The key cannot be empty'],
+			[true, 'foo', '', '', 'The key cannot be empty'],
+			[true, 'foo', 'test', '', 'Supported keys are displayname, email'],
+			[true, 'foo', 'displayname', '', 'The value cannot be empty'],
+			[true, 'foo', 'email', '', 'The value cannot be empty'],
```
Comment (sharidas) [inline tests/Core/Command/User/ModifyTest.php]: When the value is empty or null deliver the message as shown in the last index of the array.

File: tests/Core/Command/User/ModifyTest.php [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+		$this->consoleInput = $this->createMock(InputInterface::class);
+		$this->consoleOutput = $this->createMock(OutputInterface::class);
+	}
+
+	public function providerValidateInputMethod() {
+		return [
+			[false, 'foo', '', '', 'The user foo does not exist'],
+			[true, '', '', '', 'The user  does not exist'],
+			[true, null, '', '', 'The user  does not exist'],
+			[true, 'foo', null, '', 'The key cannot be empty'],
```
Comment (sharidas) [inline tests/Core/Command/User/ModifyTest.php]: When key is empty or null case addressed here.

**Review by PVince81 — APPROVED** [review]
:+1: please backport

**Comment by mmattel** [comment]
@sharidas pls add a documentation PR to add the changes to the occ command

**Comment by sharidas** [comment]
Backport PR: https://github.com/owncloud/core/pull/30652

**Comment by sharidas** [comment]
Here is the documentation ticket : https://github.com/owncloud/documentation/issues/3853
